dataprofiler/data_readers/data_utils.py

```diff
@@ -312,61 +315,71 @@ def reservoir(
 
     kinv = 1 / sample_nrows
     W = 1.0
+    rng = random.Random(x=settings._seed)
+    if "DATAPROFILER_SEED" in os.environ and settings._seed is None:
+        seed = os.environ.get("DATAPROFILER_SEED")
+        try:
+            rng = random.Random(int(seed))
+        except ValueError:
+            warnings.warn("Seed should be an integer", RuntimeWarning)
+
     while True:
-        W *= random() ** kinv
+        W *= rng.random() ** kinv
         # random() < 1.0 but random() ** kinv might not be
         # W == 1.0 implies "infinite" skips
         if W == 1.0:
             break
         # skip is geometrically distributed with parameter W
-        skip = floor( log(random())/log1p(-W) )
+        skip = floor(log(rng.random()) / log1p(-W))
         try:
-            newval = next(islice(iterator, skip, skip+1))
+            newval = next(islice(iterator, skip, skip + 1))
         except StopIteration:
             break
         # Append new, replace old with dummy, and keep track of order
-        remove_index = randrange(k)
+        remove_index = rng.randrange(sample_nrows)
         values[indices[remove_index]] = None
         indices[remove_index] = len(values)
         values.append(newval)
 
     values = [values[indices[i]] for i in irange]
-    _shuffle(values)
     return values
 
 
-def rsample(
-        file_path: Union[str, TextIOWrapper],
-        args: dict
-) -> pd.DataFrame:
+def rsample(file_path: TextIOWrapper, sample_nrows: int, args: dict) -> pd.DataFrame:
     """
-    Implement Reservoir Sampling to sample n rows out of a total of M rows in the csv file.
-    This allows sampling without loading the entire file in memory.
+    Implement Reservoir Sampling to sample n rows out of a total of M rows.
+
     :param file_path: path of the csv file to be read in
-    :type file: TextIOWrapper or string
+    :type file: TextIOWrapper
     :param args: options to read the csv file
     :type args: dict
     """
-
-    headers = []
-    header = args['header']
-    if not header:
+    head = None
+    header = args["header"]
+    if header is None:
         result = reservoir(file_path, sample_nrows)
+        fo = pd.read_csv(
+            StringIO("".join([i if (i[-1] == "\n") else i + "\n" for i in result])),
```

Comment:
@JGSweets the random sampling of the very last line of the csv file would cause an issue with reading because it would not contain the newline character at the end. So this join method now handles that specific case.

Reply:
good find!